src/utils/thread.zig

```diff
@@ -217,32 +220,74 @@ pub fn HomogeneousThreadPool(comptime TaskType: type) type {
             self.pool.deinit();
         }
 
-        pub fn schedule(self: *Self, typed_task: TaskType) Allocator.Error!void {
-            const result = try self.results.addOne();
-            var task = try self.tasks.addOne();
+        pub fn schedule(self: *Self, typed_task: TaskType) void {
+            // NOTE: this breaks other pre-scheduled tasks on re-allocs so we dont
+            // allow re-allocations
+            const result = self.results.addOneAssumeCapacity();
+            var task = self.tasks.addOneAssumeCapacity();
             task.* = .{ .typed_task = typed_task, .result = result };
             self.pool.schedule(Batch.from(&task.pool_task));
         }
 
         /// blocks until all tasks are complete
         /// returns a list of any results for tasks that did not have a pointer provided
-        pub fn join(self: *Self) std.ArrayList(TaskResult) {
+        pub fn join(self: *Self) !std.ArrayList(TaskResult) {
             for (self.tasks.items) |*task| task.join();
             const results = self.results;
-            self.results = std.ArrayList(TaskResult).init(self.allocator);
+            self.results = try std.ArrayList(TaskResult).initCapacity(self.allocator, self.tasks.capacity);
             self.tasks.clearRetainingCapacity();
             return results;
         }
 
         /// Like join, but it returns an error if any tasks failed, and otherwise discards task output.
         pub fn joinFallible(self: *Self) !void {
-            const results = self.join();
+            const results = try self.join();
             for (results.items) |result| try result;
```

Comment:
May be worth documenting that this short-circuits up to the first error.

Reply:
[ea137f5](https://github.com/Syndica/sig/pull/522/commits/ea137f5d3ee3252bbe8fc2e625d1129e5c310da9)